lib/Semantics/check-case.cpp

```diff
@@ -0,0 +1,244 @@
+                upper = lower;
+              },
+              [&](const parser::CaseValueRange::Range &x) {
+                if (x.lower) {
+                  lower = values.GetValue(*x.lower);
+                }
+                if (x.upper) {
+                  upper = values.GetValue(*x.upper);
+                }
+              },
+          },
+          range.u);
+    }
+    std::string AsFortran() const {
+      std::string result;
+      {
+        llvm::raw_string_ostream bs{result};
+        if (lower) {
+          evaluate::Constant<T>{*lower}.AsFortran(bs << '(');
+          if (!upper) {
+            bs << ':';
+          } else if (*lower != *upper) {
+            evaluate::Constant<T>{*upper}.AsFortran(bs << ':');
+          }
+          bs << ')';
+        } else if (upper) {
+          evaluate::Constant<T>{*upper}.AsFortran(bs << "(:") << ')';
+        } else {
+          bs << "DEFAULT";
+        }
+      }
+      return result;
+    }
+
+    const parser::Statement<parser::CaseStmt> &stmt;
+    std::optional<Value> lower, upper;
+  };
+
+  // Defines a comparator for use with std::list<>::sort().
+  // Returns true if and only if the highest value in range x is less
+  // than the least value in range y.  The DEFAULT case is arbitrarily
+  // defined to be less than all others.  When two ranges overlap,
+  // neither is less than the other.
+  struct Comparator {
+    bool operator()(const Case &x, const Case &y) const {
+      if (x.IsDefault()) {
+        return !y.IsDefault();
+      } else {
+        return x.upper && y.lower && *x.upper < *y.lower;
+      }
+    }
+  };
+
+  // This has quadratic time, but only runs in error cases
+  void ReportConflictingCases() {
+    for (auto iter{cases_.begin()}; iter != cases_.end(); ++iter) {
+      parser::Message *msg{nullptr};
+      for (auto p{cases_.begin()}; p != cases_.end(); ++p) {
+        if (p->stmt.source.begin() < iter->stmt.source.begin() &&
+            !Comparator{}(*p, *iter) && !Comparator{}(*iter, *p)) {
+          if (!msg) {
+            msg = &context_.Say(iter->stmt.source,
+                "CASE %s conflicts with previous cases"_err_en_US,
+                iter->AsFortran());
+          }
+          msg->Attach(
+              p->stmt.source, "Conflicting CASE %s"_en_US, p->AsFortran());
+        }
+      }
+    }
+  }
+
+  SemanticsContext &context_;
+  const evaluate::DynamicType &caseExprType_;
+  std::list<Case> cases_;
+  bool hasErrors_{false};
+};
+
+void CaseChecker::Enter(const parser::CaseConstruct &construct) {
+  const auto &selectCaseStmt{
+      std::get<parser::Statement<parser::SelectCaseStmt>>(construct.t)};
+  const auto &selectCase{selectCaseStmt.statement};
+  const auto &selectExpr{
+      std::get<parser::Scalar<parser::Expr>>(selectCase.t).thing};
+  const auto *typedExpr{selectExpr.typedExpr.get()};
+  if (!typedExpr || !typedExpr->v) {
+    return; // expression semantics failed
+  }
+  const SomeExpr &expr{*typedExpr->v};
```

Comment:
This could be:
```
  const SomeExpr *expr{GetExpr(selectExpr)};
  if (!expr) {
    return; // expression semantics failed
  }
```

Reply:
Thanks.  An earlier version of this code needed to do more with `typedExpr` but now I should use `GetExpr()`.